tests/Microsoft.Identity.Test.Unit/PublicApiTests/ConfidentialClientApplicationTests.cs

```diff
@@ -2150,5 +2153,206 @@ public void ConfidentialClient_WithInvalidAuthority_ThrowsArgumentException()
+                var mismatchHash = ComputeSHA256("some-other-token");
+                AuthenticationResult result = await app.AcquireTokenForClient(TestConstants.s_scope)
+                    .WithAccessTokenSha256ToRefresh(mismatchHash)
+                    .ExecuteAsync()
+                    .ConfigureAwait(false);
+
+                // Assert
+                // We expect to reuse the cached token
+                Assert.AreEqual("access-token", result.AccessToken,
+                    "Should reuse cached token if the hash does not match the 'bad' token's hash.");
+                Assert.AreEqual(TokenSource.Cache, result.AuthenticationResultMetadata.TokenSource);
+            }
+        }
+
+        [TestMethod]
+        public async Task ForceRefreshAndAccessTokenHash_ThrowsException_Async()
+        {
+            ConfidentialClientApplication cca = ConfidentialClientApplicationBuilder.Create(TestConstants.ClientId)
+                                                            .WithClientSecret(TestConstants.ClientSecret)
+                                                            .WithExperimentalFeatures(true)
+                                                            .BuildConcrete();
+
+            // Attempt to create a request that sets both ForceRefresh and an AccessTokenHash
+            AcquireTokenForClientParameterBuilder builder = cca.AcquireTokenForClient(TestConstants.s_scope)
+                                .WithForceRefresh(true)
+                                .WithAccessTokenSha256ToRefresh("dummyHash");
+
+            // Act & Assert
+            MsalClientException ex = await AssertException.TaskThrowsAsync<MsalClientException>(() =>
+                builder.ExecuteAsync()
+            ).ConfigureAwait(false);
+
+            Assert.AreEqual(MsalError.ForceRefreshNotCompatibleWithTokenHash, ex.ErrorCode);
+            Assert.IsTrue(ex.Message.Contains(MsalErrorMessage.ForceRefreshAndTokenHasNotCompatible));            
+        }
+
+        [TestMethod]
+        public async Task AcquireTokenForClient_WithClaims_And_MatchingHash_SkipsCache_Async()
+        {
+            using (var httpManager = new MockHttpManager())
+            {
+                httpManager.AddInstanceDiscoveryMockHandler();
+
+                // 1) Create app & fetch an initial token
+                var app = ConfidentialClientApplicationBuilder
+                    .Create(TestConstants.ClientId)
+                    .WithClientSecret(TestConstants.ClientSecret)
+                    .WithHttpManager(httpManager)
+                    .WithExperimentalFeatures(true)
+                    .BuildConcrete();
+
+                httpManager.AddMockHandlerSuccessfulClientCredentialTokenResponseMessage(token: "old-token");
+                AuthenticationResult firstResult = await app.AcquireTokenForClient(TestConstants.s_scope).ExecuteAsync().ConfigureAwait(false);
+
+                Assert.AreEqual("old-token", firstResult.AccessToken);
```

Comment:
nit: Should you define "old-token" constant since it's used more than once? Same feedback to all other constants you're using.